### ISSUE TITLE
KRB_AP_REQ not exposed in the library

### ISSUE THREAD

**Issue by Hochman85** [antiduh/nsspi#25]
Hi,

I am trying to get hold of KRB_AP_REQ to use in our API authentication (Application Server specific) but can't seem to find any field that looks like KRB_AR_REQ. If that's not exposed, I'll have to build it myself.
https://www.tarlogic.com/en/blog/how-kerberos-works/

Can you please confirm the library doesn't expose that field. 
Thanks

**Comment by antiduh** [comment]
NSSPI is just a wrapper around the Win32 SSPI API. In general: if it's not possible in SSPI, it's not possible in NSSPI.

However, SSPI (and NSSPI) support Kerberos, and the message you're describing sounds like one of the core authentication cycle messages that get sent by the Kerberos security provider.

If the functionality that NSSPI provides includes secure mutual authentication with Kerberos, why do you need this message specifically? 

The whole point of the SSPI design and thus NSSPI, is you tell it which protocol you want (Kerberos, NTLM, Negotiate), what options you want, and it tells you what bytes to send without you having to bother knowing what the messages mean.

Are you trying to implement GSSAPI authentication? It's possible but not yet directly supported by NSSPI. If you are, I'd recommend you read about the GSSAPI mechanisms in the SSPI documentation.

https://docs.microsoft.com/en-us/windows/win32/secauthn/sspi

https://docs.microsoft.com/en-us/windows/win32/secauthn/sspi-kerberos-interoperability-with-gssapi

**Comment by Hochman85** [comment]
It's not really GSSAPI what I am after. All I want is: get the Kerberos token off the currently logged on user (already done thanks to NSSPI which covers the "client-KDC" journey), then using NSSPI or another library (not sure), get access to AP-REQ so I can send it from client to AP (application server). I think there's three possibilities here:
1- either NSSPI provides a mechanism to complete the client-AP journey at a high level without giving access to AP-REQ (this works for me if it was the case so I can only call something like client.AuthenticateToAP(...)
2- or, get access to AP-REQ which I can then send to AP using a POST API call - you already mentioned the field in question was not exposed in NSSPI
3- or, build AP-REQ myself using the Kerberos token I have. there's two particular methods that can perhaps help me here: client.MakeSignature(...) and client.Encrypt(...)

Can you please point me to what you think is the right next step: 1 or 3 ("2" ruled out) for client-AP journey?

Thanks for your help.

Maybe I have to combine two libraries here: SSPI and Kerberos.NET. If you could explain to me how to use Kerberos.NET.Client.ApplicationSessionContext's ApReq please, that'd narrow things down I suppose.

Thanks for your help.

**Comment by antiduh** [comment]
You're really misunderstanding the point of SSPI. Option 1 is *the entire purpose* of the SSPI system.

What software calls SSPI's InitializeSecurityContext, and what software calls SSPI's AcceptSecurityContext?

The answer to both is never the KDC. As far as you're concerned you don't even know that a KDC exists.

InitializeSecurityContext is called by your client software.
AcceptSecurityContext is called by your application server.

Go re-read the main page and documentation on SSPI. The 'authentication cycle' I describe on the main page describes how the client and application server send opaque messages to each other to establish a shared security context.

One last time: SSPI is used by the client and application service. Go make a demo app with NSSPI and come back only after you've figured out how it's meant to be used.

> If you could explain to me how to use Kerberos.NET.Client.ApplicationSessionContext's ApReq please

No. I'm not your tutor. That's not my software.
